File: disposia-engine/src/main/java/io/disposia/engine/mapper/EpisodeMapper.java
package io.disposia.engine.mapper;

import io.disposia.engine.domain.IndexField;
import io.disposia.engine.domain.Episode;
import io.disposia.engine.domain.ImmutableEpisode;
import io.disposia.engine.domain.ModifiableEpisode;
import org.apache.solr.common.SolrDocument;
import org.mapstruct.Mapper;
import org.mapstruct.MappingTarget;
import org.mapstruct.NullValueCheckStrategy;
import org.mapstruct.factory.Mappers;

import java.util.Objects;
import java.util.Optional;
import java.util.stream.Stream;

@Mapper(uses = {PodcastMapper.class, ChapterMapper.class, DateMapper.class},
        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS)
public interface EpisodeMapper {

    EpisodeMapper INSTANCE = Mappers.getMapper( EpisodeMapper.class );

    default ModifiableEpisode toModifiable(Episode episode) {
        return Optional
            .ofNullable(episode)
            .map(e -> {
                if (e instanceof ModifiableEpisode) {
                    return (ModifiableEpisode) e;
                } else {
                    return new ModifiableEpisode().from(e);
                }})
            .orElse(null);
    }

    default ImmutableEpisode toImmutable(Episode episode) {
        return Optional
            .ofNullable(episode)
            .map(e -> {
                if (e instanceof ImmutableEpisode) {
                    return (ImmutableEpisode) e;
                } else {
                    return ((ModifiableEpisode) e).toImmutable();
                }})
            .orElse(null);
    }

    ModifiableEpisode update(Episode src, @MappingTarget ModifiableEpisode target);

    default ModifiableEpisode update(Episode src, @MappingTarget Episode target) {
        return Optional
            .ofNullable(target)
            .map(t -> {
                if (t instanceof ModifiableEpisode) {
                    return (ModifiableEpisode) t;
                } else {
                    return new ModifiableEpisode().from(t);
                }})
            .map(t -> update(src, t))
            .orElse(null);
    }

    default ImmutableEpisode updateImmutable(Episode src, @MappingTarget Episode target) {
        return Optional
            .ofNullable(target)
            .map(t -> {
                if (t instanceof ModifiableEpisode) {
                    return (ModifiableEpisode) t;
                } else {
                    return new ModifiableEpisode().from(t);
                }})
            .map(t -> update(src, t))
            .map(ModifiableEpisode::toImmutable)
            .orElse(null);
    }

    default ImmutableEpisode toImmutable(org.apache.lucene.document.Document doc) {
        return Optional
            .ofNullable(doc)
            .map(d -> ImmutableEpisode.builder()
                .setId(d.get(IndexField.ID))
                //.setExo(d.get(IndexField.EXO))
                .setTitle(d.get(IndexField.TITLE))
                .setLink(d.get(IndexField.LINK))
                .setPubDate(Optional
                    .ofNullable(d.get(IndexField.PUB_DATE))
                    .map(DateMapper.INSTANCE::asLocalDateTime)
                    .orElse(null))
                .setDescription(Stream
                    .of(d.get(IndexField.ITUNES_SUMMARY), d.get(IndexField.DESCRIPTION))
                    .filter(Objects::nonNull)
                    .findFirst()
                    .orElse(null))
                .setImage(d.get(IndexField.ITUNES_IMAGE))
                .setItunesDuration(d.get(IndexField.ITUNES_DURATION))
                .setPodcastTitle(d.get(IndexField.PODCAST_TITLE))
                .create())
            .orElse(null);
    }

    default ImmutableEpisode toImmutable(SolrDocument doc) {
        return Optional
            .ofNullable(doc)
            .map(d -> ImmutableEpisode.builder()
                .setId(SolrFieldMapper.INSTANCE.stringOrNull(d, IndexField.ID))
                //.setExo(SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.EXO))
                .setTitle(SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.TITLE))
                .setLink(SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.LINK))
                .setPubDate(Optional
                    .ofNullable(SolrFieldMapper.INSTANCE.firstDateOrNull(d, IndexField.PUB_DATE))
                    .map(DateMapper.INSTANCE::asLocalDateTime)
                    .orElse(null))
                .setDescription(Stream
                    .of(SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.ITUNES_SUMMARY), SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.DESCRIPTION))
                    .filter(Objects::nonNull)
                    .findFirst()
                    .orElse(null))
                .setImage(SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.ITUNES_IMAGE))
                .setItunesDuration(SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.ITUNES_DURATION))
                .setPodcastTitle(SolrFieldMapper.INSTANCE.firstStringOrNull(d, IndexField.PODCAST_TITLE))
                .create())
            .orElse(null);
    }

}

File: disposia-engine/src/main/scala/io/hemin/engine/util/mapper/EpisodeMapper.scala
package io.hemin.engine.util.mapper

import io.hemin.engine.domain.info.EpisodeItunesInfo
import io.hemin.engine.domain.{Episode, IndexDoc}
import io.hemin.engine.domain.IndexField
import org.apache.solr.common.SolrDocument

object EpisodeMapper {

  def toEpisode(src: IndexDoc): Episode = Option(src)
    .map{ s =>
      Episode(
        id          = s.id,
        title       = s.title,
        link        = s.link,
        description = s.description,
        pubDate     = s.pubDate,
        image       = s.image,
        itunes = EpisodeItunesInfo(
          author  = s.itunesAuthor,
          summary = s.itunesSummary,
          //duration = s.itunesDuration,
        )
      )
    }
    .orNull


  def toEpisode(src: org.apache.lucene.document.Document): Episode = Option(src)
    .map { s =>
      Episode(
        id           = LuceneMapper.get(s, IndexField.ID),
        title        = LuceneMapper.get(s, IndexField.TITLE),
        podcastTitle = LuceneMapper.get(s, IndexField.PODCAST_TITLE),
        link         = LuceneMapper.get(s, IndexField.LINK),
        pubDate      = DateMapper.asLocalDateTime(s.get(IndexField.PUB_DATE)),
        description  = LuceneMapper.get(s, IndexField.DESCRIPTION),
        image        = LuceneMapper.get(s, IndexField.ITUNES_IMAGE),
        itunes = EpisodeItunesInfo(
          author   = LuceneMapper.get(s, IndexField.ITUNES_AUTHOR),
          summary  = LuceneMapper.get(s, IndexField.ITUNES_SUMMARY),
          duration = LuceneMapper.get(s, IndexField.ITUNES_DURATION),
        )
      )
    }.orNull


  def toEpisode(src: SolrDocument): Episode = Option(src)
    .map { s =>
      Episode(
        id           = SolrMapper.firstStringMatch(s, IndexField.ID),
        title        = SolrMapper.firstStringMatch(s, IndexField.TITLE),
        podcastTitle = SolrMapper.firstStringMatch(s, IndexField.PODCAST_TITLE),
        link         = SolrMapper.firstStringMatch(s, IndexField.LINK),
        pubDate      = SolrMapper.firstDateMatch(s, IndexField.PUB_DATE).flatMap(x => DateMapper.asLocalDateTime(x)),
        description  = SolrMapper.firstStringMatch(s, IndexField.DESCRIPTION),
        image        = SolrMapper.firstStringMatch(s, IndexField.ITUNES_IMAGE),
        itunes = EpisodeItunesInfo(
          author   = SolrMapper.firstStringMatch(s, IndexField.ITUNES_AUTHOR),
          summary  = SolrMapper.firstStringMatch(s, IndexField.ITUNES_SUMMARY),
          duration = SolrMapper.firstStringMatch(s, IndexField.ITUNES_DURATION),
        )
      )
    }.orNull


}
